### ISSUE TITLE
Support of ThreadSafeReference 

### ISSUE THREAD

**Issue by AleksanderMielczarek** [realm/realm-java#4059]
Hi,

I have a question about ThreadSafeReference. This feature appeared recently for Objective-C and Swift: https://realm.io/news/obj-c-swift-2-2-thread-safe-reference-sort-properties-relationships/.
Is this feature planned also for Java in any of future releases?

Best regards
Aleksander Mielczarek

**Comment by Zhuinden** [comment]
I believe it depends on https://github.com/realm/realm-java/pull/3834 but it'll be pretty cool when it arrives 😄

**Comment by beeender** [comment]
Yup, surely we will support this in java!
One issue currently in my mind is, to support this, we will have to pin a `Realm` version to the `ThreadSafeRealmObject` (must be a better name than this!), unlike cocoa, we rely on GC to control the life cycle of the object, which means the pinned version will not be released until that thread-safe-object get GCed. potentially increase the Realm file size if there are many leaked thread-safe-objects belong to different Realm versions.

But let's see, it will be a very cool and important feature!

**Comment by Zhuinden** [comment]
@beeender that sounds oddly reminiscient for https://github.com/realm/realm-java/issues/1208

**Comment by cmelchior** [comment]
Yes, depending on how the API ends up looking this might very well replace #1208, at this point we need to spec out the specifics though

**Comment by Zhuinden** [comment]
The principle behind this doesn't look too difficult though, it seems this is for passing **1** object between two threads, so that you don't need to get the ID and requery the object on the other thread anymore.

I'm probably missing the complexity at initial glance, though. This will be useful!

**Comment by Zhuinden** [comment]
What's blocking this one? Based on the cocoa docs, it seems to be just a wrapper around a proxy object which allows you to "pass it" across the thread so that you don't need to requery by id.
